diff --git a/scripts/cros_setup_toolchains.py b/scripts/cros_setup_toolchains.py
index 14825c119..4074dab5a 100644
--- a/scripts/cros_setup_toolchains.py
+++ b/scripts/cros_setup_toolchains.py
@@ -84,6 +84,7 @@ TARGET_GO_ENABLED = (
     'armv7a-cros-linux-gnueabi',
     'armv7a-cros-linux-gnueabihf',
     'aarch64-cros-linux-gnu',
+    'riscv64-cros-linux-gnu',
 )
 CROSSDEV_GO_ARGS = ['--ex-pkg', 'dev-lang/go']
 
@@ -93,6 +94,7 @@ TARGET_COMPILER_RT_ENABLED = (
     'armv7a-cros-linux-gnueabihf',
     'aarch64-cros-linux-gnu',
     'armv7m-cros-eabi',
+    'riscv64-cros-linux-gnu',
 )
 CROSSDEV_COMPILER_RT_ARGS = ['--ex-pkg', 'sys-libs/compiler-rt']
 
@@ -102,6 +104,7 @@ TARGET_LLVM_PKGS_ENABLED = (
     'aarch64-cros-linux-gnu',
     'i686-pc-linux-gnu',
     'x86_64-cros-linux-gnu',
+    'riscv64-cros-linux-gnu',
 )
 
 LLVM_PKGS_TABLE = {
